roboplot/imgproc/image_analysis.py

```diff
@@ -22,414 +27,492 @@ class Direction(enum.IntEnum):
+                break
 
-            indices = np.asarray([x_indices[index], y_indices[index]])
-            error = InterpolateAverages.error(line_normal, origin, indices)
+                last_centroid = Centroid.INVALID_NO_WHITE
 
-            if error > max_error:
-                return index
+    # Return the list of average indices and the scan direction for the next picture taken.
+    return indices, turn_to_next_scan
 
-        return -1
 
-    @staticmethod
-    def error(line_normal, line_origin, indices):
+def approximate_with_line(indices):
+    """
+    Approximates gradient of line assuming that the line goes through the start point
+    Args:
+        indices: The points to be approximated.
+
+    Returns:
+        The gradient and constant of the approximation line.
+    """
+
+    # Indices are ordered y, x in list.
+    x_translated = np.array(indices)[:, 1] - indices[0][1]
+    y_translated = np.array(indices)[:, 0] - indices[0][0]
+    y_translated = y_translated[:, np.newaxis]
+    a, _, _, _ = np.linalg.lstsq(y_translated, x_translated)
 
-        if indices[0] != -1:
-            error = abs(np.dot(line_normal, indices - line_origin))
-        else:
-            error = 0.0
+    c = indices[0][1] - indices[0][0] * a[0]
+    return a[0], c
+
+
+def error_from_line(line, indices, max_error):
+    """
+        This functions finds the first index along the line where the error is greater than the max error bound.
+    Args:
+        line: The coefficients of the line approximation.
+        indices: The indices being approximated.
+        max_error: The max error allowed from the line
+
+    Returns:
+        The indices of the point that first breaks the maximum error.
+    """
+    if line[0]:
+        line_normal = np.array([1, -1 / line[0]])
+        line_normal /= math.sqrt(1 + 1 / (line[0] * line[0]))
+    else:
+        line_normal = np.array([0, 1])
+
+    origin = np.asarray([0, line[1]])
+
+    # Note that we don't care what the error is to the first 2 indices as we are forcing the start position as
+    # we want the line to be continuous.
+    for index in range(2, len(indices)):
+        error = compute_error(line_normal, origin, indices[index])
+        if error > max_error:
+            return index
+    return -1
+
+
+def compute_error(line_normal, line_origin, indices):
+    if indices[0] != -1:
+        error = abs(np.dot(line_normal, indices - line_origin))
+    else:
+        error = 0.0
+    return error
+
+
+def rotate(origin, point, angle):
+    """
+    Rotate a point counterclockwise by a given angle around a given origin.
+
+    The angle should be given in radians.
+    """
+    ox, oy = origin
+    px, py = point
+
+    sin_angle = math.sin(angle)
+    cos_angle = math.cos(angle)
+
+    qx = ox + cos_angle * (px - ox) - sin_angle * (py - oy)
+    qy = oy + sin_angle * (px - ox) + cos_angle * (py - oy)
+    return [int(qx), int(qy)]
+
+
+def approximate_path(pixel_indices):
```

Comment:
This is a copy and paste from the Internet.  So I don't understand it myself. But it works in the tests.